### ISSUE TITLE
4.2.0 broken (typescript + webpack):

### ISSUE THREAD

**Issue by nbelyh** [janl/mustache.js#786]
// foo.ts:
```
import * as mustache from "mustache"

// runtime error: Uncaught TypeError: mustache__WEBPACK_IMPORTED_MODULE_2__.render is not a function???
mustache.render('foo', {}); 
``
Reverting to 4.1.0 fixes the problem.

The

**Comment by nbelyh** [comment]
Anyone? The issue won't fix itself 😄

**Comment by phillipj** [comment]
Hi @nbelyh,

thanks for reporting! 👍🏻 

Any chance you could make that reproduction available as a github.com repository? ..I've gotten the habit of not opening .zip archives from the internetz over the years 😸

**Comment by nbelyh** [comment]
@phillipj thank you for response. Here you go:

https://github.com/nbelyh/repro-mustache-402

> npm install
> npm start (just launches webpack)

Open the file index.html => error in the browser console (from the description)

> npm install mustache@4.1.0
> npm start

The page opens fine (no errors)

**Comment by nbelyh** [comment]
@phillipj forgot to make the repo public 🤦 - now it should be

**Comment by phillipj** [comment]
Sorry for the late reply, and thanks for reporting!

You've definitively hit an unexpected side effect that we did not consider when introducing the package.json/`exports` field in https://github.com/janl/mustache.js/pull/773. Those changes seems to make webpack choose the .mjs with ES Modules syntax instead of .js, and the consequence is slighty different semantics when importing mustache.js.

Gotta dive deeper and compare pros/cons on plausible solutions going forward, in the meantime if you want to stay on mustache.js 4.2.0, the following does the trick and makes sense having the mustache.js' source code in mind:

```js
import mustache from "mustache"
```

**Comment by nbelyh** [comment]
@phillipj Thank you for coming to this!
Unfortunately, this import syntax requires esModuleInterop=true in tsconfig. It is not always possible to have it like this, as it may conflict with other imports.

**Comment by phillipj** [comment]
> Unfortunately, this import syntax requires esModuleInterop=true in tsconfig (this issue is about typescript). It is not always possible to have it like this, as it may conflict with other imports.

Huh, that's weird, as I was able to use the `import` I suggested in that reproduction repository of yours, without any other changes.

> For now I've decided to stay with 4.1.0...

Alrighty :+1:

**Comment by nbelyh** [comment]
Yes you are right, it works in this repository without this directive. I faced the problem with this import in a working project, will check it there and hopefully come back with more details. Maybe typescript version...

**Comment by ricardodev2015** [comment]
Reverting to 4.1.0, until 4.2.0 is solved. I'm getting this error in Chrome:

![Mustache4 2 0-Error](https://user-images.githubusercontent.com/10821755/150396020-4d550b0c-94db-4783-bc98-627bb5dd39fc.png)

**Comment by deltamualpha** [comment]
yes, this this a major breaking change for those of us using mustache in a very simple `<script type="text/javascript" src="js/mustache.4.1.0.js"></script>` sort of way. The 4.2.0 file is invalid unless imported as `type="module"`, which, while the "modern" way to do native module support, is definitely a forced change.

**Comment by phillipj** [comment]
Hi folks,

my lack of responses and alternatives to fix this properly is a result of me not having the time to do much open source in my spare time these days. I'm obviously sorry for the frustration and wasted hours this has caused for you.

Your reports are certainly valuable and needed to ensure we consider more usage scenarios than what we obviously had in mind when doing the 4.2.0 release -- we honestly tried our best to consider everything, but failed.

Now that we do have a few added scenarios to consider, it feels the most valuable contribution going forward would be ideas of what is wrong and what could be done differently. My capacity isn't going to change in the near future, so in the spirit of open source, this needs to be fixed by collaborating around thoughts and/or concrete changes to the code or build process.

_Bottom line:_ help mustache.js get this right please?

**Comment by deltamualpha** [comment]
@phillipj totally understandable. I'm no javascript wiz, but I'll take a look at the repo and its build process when I have a chance and see what I can come up with. It might be that mustache.js needs to provide a couple of variants, built for specific tool-chains.
